File: frontend/src/api/modules/accounts.ts
import type {
  AccountInviteListResponse,
  AccountInviteResponse,
  AccountListResponse,
  AccountResponse,
  AccountUsersResponse,
  InviteStatusResponse,
  PermissionListResponse,
  RebindUserResponse,
  RoleDeleteResponse,
  RoleInfoResponse,
  RoleListResponse,
  UserRoleUpdateResponse,
} from '../generated'
import api from '../index'

export default {
  me: () => api.get<AccountResponse>('/api/accounts/me'),

  list: () => api.get<AccountListResponse>('/api/accounts'),

  create: (data: { code: string, name: string }) =>
    api.post<AccountResponse>('/api/accounts', data),

  updateStatus: (accountId: number, status: 'active' | 'disabled') =>
    api.put<AccountResponse>(`/api/accounts/${accountId}/status`, { status }),

  listUsers: (accountId: number) =>
    api.get<AccountUsersResponse>(`/api/accounts/${accountId}/users`),

  updateUserRole: (accountId: number, userId: number, role: string) =>
    api.put<UserRoleUpdateResponse>(`/api/accounts/${accountId}/users/${userId}/role`, { role }),

  updateUserRoles: (accountId: number, userId: number, role_codes: string[]) =>
    api.put<UserRoleUpdateResponse>(`/api/accounts/${accountId}/users/${userId}/roles`, { role_codes }),

  rebindUser: (accountId: number, userId: number, migrate_data: boolean) =>
    api.put<RebindUserResponse>(`/api/accounts/${accountId}/users/${userId}`, { migrate_data }),

  createInvite: (accountId: number, expires_in_hours: number) =>
    api.post<AccountInviteResponse>(`/api/accounts/${accountId}/invites`, { expires_in_hours }),

  listInvites: (accountId: number) =>
    api.get<AccountInviteListResponse>(`/api/accounts/${accountId}/invites`),

  revokeInvite: (inviteId: number, reason = '') =>
    api.put<InviteStatusResponse>(`/api/accounts/invites/${inviteId}/revoke`, { reason }),

  permissions: () => api.get<PermissionListResponse>('/api/accounts/permissions'),

  roles: (accountId?: number) => api.get<RoleListResponse>('/api/accounts/roles', {
    params: accountId ? { account_id: accountId } : undefined,
  }),

  createRole: (accountId: number, data: { code: string, name: string, description: string, permission_codes: string[] }) =>
    api.post<RoleInfoResponse>(`/api/accounts/${accountId}/roles`, data),

  updateRole: (
    accountId: number,
    roleId: number,
    data: { name: string, description: string, status: string, permission_codes: string[] },
  ) => api.put<RoleInfoResponse>(`/api/accounts/${accountId}/roles/${roleId}`, data),

  deleteRole: (accountId: number, roleId: number) =>
    api.delete<RoleDeleteResponse>(`/api/accounts/${accountId}/roles/${roleId}`),
}

File: backend/app/schemas/accounts.py
from __future__ import annotations

from typing import Any

from pydantic import Field

from app.schemas.common import ApiModel, ListResponse, WarningMixin


class AccountResponse(ApiModel):
    id: int
    code: str
    name: str
    status: str
    user_count: int | None = None
    created_at: str | None = None
    updated_at: str | None = None


class AccountListResponse(ListResponse[AccountResponse]):
    pass


class PermissionInfoResponse(ApiModel):
    id: int
    code: str
    name: str
    description: str
    is_system: bool


class PermissionListResponse(ListResponse[PermissionInfoResponse]):
    pass


class RoleInfoResponse(ApiModel):
    id: int
    account_id: int
    code: str
    role: str
    name: str
    description: str
    status: str
    is_system: bool
    permissions: list[str] = Field(default_factory=list)


class RoleListResponse(ListResponse[RoleInfoResponse]):
    pass


class UserRoleSummaryResponse(ApiModel):
    id: int
    code: str
    name: str
    is_system: bool


class AccountUserResponse(ApiModel):
    id: int
    username: str
    display_name: str
    department: str
    role: str
    role_codes: list[str] = Field(default_factory=list)
    roles: list[UserRoleSummaryResponse] = Field(default_factory=list)
    created_at: str | None = None


class AccountUsersResponse(ApiModel):
    account: AccountResponse
    items: list[AccountUserResponse] = Field(default_factory=list)
    total: int


class AccountInviteResponse(ApiModel):
    id: int
    account_id: int | None = None
    status: str
    created_by: int | None = None
    used_by: int | None = None
    created_at: str | None = None
    used_at: str | None = None
    expires_at: str | None = None
    invite_id: int | None = None
    invite_code: str | None = None


class AccountInviteListResponse(ListResponse[AccountInviteResponse]):
    pass


class InviteStatusResponse(ApiModel):
    id: int
    status: str


class RoleDeleteResponse(ApiModel):
    id: int
    deleted: bool


class RebindUserResponse(WarningMixin):
    user_id: int
    old_account_id: int
    new_account_id: int
    migrated: bool
    migrate_data: bool
    counts: dict[str, int] = Field(default_factory=dict)


class UserRoleUpdateResponse(ApiModel):
    id: int
    role: str
    role_codes: list[str] = Field(default_factory=list)
